Add DevTools Airlock explainer (#466)
MicrosoftEdge/MSEdgeExplainers: DevTools/Airlock/explainer.md

## Goals

  1. Provide Airlock as a new tool in the Microsoft Edge DevTools to better debug DOM node memory leaks

We're eager to hear your feedback about Airlock and the current state of memory tooling in the Microsoft Edge DevTools. Let us know what you think by filing an [issue on this repo](https://github.com/MicrosoftEdge/MSEdgeExplainers/issues/new?labels=DevTools&template=devtools.md&title=%5BDevTools%5D+%3CTITLE+HERE%3E) or on [Twitter](https://twitter.com/EdgeDevTools).

## Usage

**Note:** Airlock is not currently available in the Microsoft Edge DevTools. The [Changelog section](#changelog) will be updated as soon as a version of Microsoft Edge with Airlock is available.

### What is a DOM node memory leak?

> A DOM node memory leak is when a DOM node that is removed from the DOM tree cannot be garbage collected because it is still referenced by the page's JavaScript. Nodes that fall into this category are considered "detached". Depending on the parents and children of the node that causes the leak, a DOM node memory leak can span from just a few nodes all the way to massive trees of hundreds or thousands of elements. This can also be a common occurrence for React apps because React maintains a virtual representation of the DOM in JavaScript.

Reviewer: ```suggestion
A DOM node memory leak occurs when a DOM node that is detached from the DOM tree cannot be garbage collected because a reference to it still exists. Nodes that fall into this category are considered "detached". A detached node's parent and child references can cause a single node memory leak to span just a few nodes all the way to massive trees of hundreds or thousands of elements. This can also be a common occurrence for React apps because React maintains a virtual representation of the DOM in JavaScript.
```

<!-- it would be a good idea to host Brian's sample in the EdgeDevToolsSamples repo and point to it from here -->

### Airlock

> Airlock displays all of the detached DOM nodes and the size of the tree for each detached node. By triggering GC and querying for the detached nodes in Airlock, you are able to find all the detached nodes that are still allocated in memory. In the following figure, Airlock displays 3 detached nodes, meaning 3 trees of 5 nodes each are still being retained in memory despite being removed from the DOM tree.

Reviewer: ```suggestion
The Airlock DevTools browser extension hooks onto `document.createElement` calls to track the lifecycle of DOM nodes on the target page. Airlock can display all of the detached DOM nodes and the size of the connected tree for each detached node. After triggering garbage collection and querying for detached nodes in Airlock, nodes that were not garbage collected will remain in view.

In the following figure, Airlock displays 3 detached nodes, meaning 3 trees of 5 nodes each are still being retained in memory despite being removed from the DOM tree. (Although each tree only shows three elements, there are two text nodes present for a total of five).
```

> ![Airlock](airlock.png)

Reviewer: ```suggestion
![Airlock](airlock.png)

Because a **single** detached node can result in **all** related tree nodes being leaked (via references such as `parentNode` or `childNodes`), it can be difficult to determine which specific node within a tree caused the tree to leak.

The **Detach roots** feature destructively disconnects nodes within a tree so they no longer reference one another. When garbage collection is run again, nodes which had no more references to them will be freed, leaving only individual leaky nodes nodes in place. The Memory heap snapshot tool can then be used to identify the offending retaining references to the leaky nodes.
```

<!-- this image should be updated when we have mock-ups from design on how we want Airlock to look -->

The following figure documents the current UI for the Airlock DevTools extension. **Note** that this UI will be updated to match the look and feel of the Microsoft Edge DevTools when Airlock is integrated as its own tool.

![Airlock UI markup](airlock_ui_markup.png)

> When Airlock is integrated as its own tool in the Microsoft Edge DevTools, it will provide the following features:

Reviewer: ```suggestion
## Airlock + DevTools Integration

By integrating Airlock as a native DevTools tool, we can streamline the Airlock user experience by automating the manual steps required to diagnose leaky detached nodes while also folding Airlock's capabilities into related DevTools tools like Elements or Memory. The following features are planned:
```
